modules/azure/client_factory.go

```diff
@@ -0,0 +1,147 @@
+	ManagedClustersClientType
+)
+
+// ClientFactory describes the methods available on client factory implementatoins
+type ClientFactory interface {
+	// GetClientE returns a client instance based on the ClientType passed, or optionally an error.
+	GetClientE(clientType ClientType, subscriptionID string) (interface{}, error)
+
+	// GetClientBaseURIE returns the configured BaseURI for the given client, based on the passed ClientType.
+	GetClientBaseURIE(clientType ClientType, client interface{}) (string, error)
+}
+
+// multiEnvClientFactory is used to coordinate handing out properly configured Azure SDK clients
+// that are properly setup for use with Public or Sovereign clouds (depending on configuration)
+type multiEnvClientFactory struct{}
+
+// NewClientFactory returns a new multi-environment client factory
+func NewClientFactory() ClientFactory {
+	return &multiEnvClientFactory{}
+}
+
+// GetClientE returns a client instance based on the ClientType passed, or optionally an error.
+func (factory *multiEnvClientFactory) GetClientE(clientType ClientType, subscriptionID string) (interface{}, error) {
+	// Validate Azure subscription ID
+	subscriptionID, err := getTargetAzureSubscription(subscriptionID)
+	if err != nil {
+		return nil, err
+	}
+
+	// Lookup environment URI
+	baseURI, err := factory.getEnvironmentBaseURIE()
+	if err != nil {
+		return nil, err
+	}
+
+	// Create correct client based on type passed
+	switch clientType {
+	case SubscriptionsClientType:
+		return subscriptions.NewClientWithBaseURI(baseURI), nil
+	case VirtualMachinesClientType:
+		return compute.NewVirtualMachinesClientWithBaseURI(baseURI, subscriptionID), nil
+	case ManagedClustersClientType:
+		return containerservice.NewManagedClustersClientWithBaseURI(baseURI, subscriptionID), nil
+	}
+
+	// If nothing matched, this is an error
+	return nil, fmt.Errorf("Unknown client type %s", string(clientType))
+}
+
+// GetClientBaseURIE returns the configured BaseURI for the given client, based on the passed ClientType.
+func (factory *multiEnvClientFactory) GetClientBaseURIE(clientType ClientType, client interface{}) (string, error) {
+	// Create correct client based on type passed
+	switch clientType {
+	case SubscriptionsClientType:
+		return client.(subscriptions.Client).BaseURI, nil
+	case VirtualMachinesClientType:
+		return client.(compute.VirtualMachinesClient).BaseURI, nil
+	case ManagedClustersClientType:
+		return client.(containerservice.ManagedClustersClient).BaseURI, nil
+	}
+
+	// If no matches, client was unknown
+	return "", fmt.Errorf("Unknown client type %s", string(clientType))
+}
+
+// getDefaultEnvironmentName returns either a configured Azure environment name, or the public default
+func (factory *multiEnvClientFactory) getDefaultEnvironmentName() string {
+	envName, exists := os.LookupEnv(AzureEnvironmentEnvName)
+
+	if !exists || envName == "" {
+		envName = autorestAzure.PublicCloud.Name
+	}
+
+	return envName
```

Comment:
NIT: slightly more functional programming / immutable approach, with simpler conditional:

```suggestion
	if exists && len(envName) > 0 {
		return envName
	}
	
	return autorestAzure.PublicCloud.Name
```

Reply:
Done - thanks for the suggestion!